src/main/scala/org/stormenroute/mecha/package.scala

```diff
@@ -432,4 +432,66 @@ package object mecha {
     def default(v: =>T): Input.Query[T] = Input.default(query)(v)
   }
 
+  private[mecha] object GitIgnore {
+    import java.nio.charset.StandardCharsets._
+    import java.nio.file.{FileSystems, Paths, Files, Path}
+    import java.nio.file.StandardOpenOption._
+    import scala.collection.JavaConversions._
+
+    val log = MechaLog.Println
+
+    def ignore(ignoreFilePath: Path, toAdd: String) = {
+      log.info(s"Adding '$toAdd' to '${ignoreFilePath.toString}'...")
+      writeFile(ignoreFilePath, add(readFile(ignoreFilePath), toAdd))
+    }
+
+    private def readFile(file: Path): Seq[String] = {
+      Files.readAllLines(file, UTF_8)
+    }
+
+    private def writeFile(file: Path, content: Seq[String]) = {
+      Files.write(file, content, UTF_8, WRITE, CREATE)
+    }
+
+    private def add(existing: Seq[String], `new`: String): Seq[String] = {
+      val patterns = existing.map(line).collect({case p: Pattern => p})
+      val alreadyIgnored = patterns.find(p => p.matches(`new`))
+      if (alreadyIgnored.isDefined) {
+        log.info(s"'${`new`}' is already ignored via pattern '${alreadyIgnored.get.p}'")
+        existing
+      }
+      else {
+        existing :+ `new`
+      }
+    }
+
+    private def line(l: String): Line = l.trim match {
+      case "" => Empty
+      case s if s.startsWith("#") => Comment(s)
+      case _ => Pattern(l)
+    }
+
+    sealed trait Line {
+      def matches(path: String): Boolean = false
+    }
+    object Empty extends Line
+    case class Comment(c: String) extends Line
+    case class Pattern(p: String) extends Line {
+      val trimmed = p.trim
+      val negative = trimmed.startsWith("!")
+      lazy val matcher = FileSystems.getDefault.getPathMatcher(s"glob:./${cleanUp(trimmed)}")
+      override def matches(dirName: String): Boolean = {
+        val path = if (!dirName.startsWith("./")) s"./$dirName" else dirName
+        val matches = matcher.matches(Paths.get(path))
+        (!negative && matches) || (negative && !matches)
```

Comment:
My understanding is that patterns starting with `!` in ignore/exclude files override blacklist, roughly speaking.

If so, won't this consider a line `alreadyIgnored` if there is a whitelist pattern that does not match the new line?
In other words, if you have this:

```
*.class
abc.class
```

And the new line that you want to insert is `foobar`, then `matches` returns `Some("!abc.class")` for `foobar`, even though `foobar` is not ignored.

You probably want to have two methods `blacklists` and `whitelists`, and conclude that something is already ignored if it `blacklists` returns non-`None`, but `whitelists` returns `None`.